### PR TITLE
Switch to using OpenSearch

### PR DESCRIPTION
**Pull request by dbason**
This pull requests switches from the deprecated Opendistro for ES to OpenSearch.  It also adds functionality to manage upgrades.

The data node upgrades will occur first and will be managed.  After that the standard kubernetes rollout mechanisms will be used for other node types.

### CODE REVIEW

**Review by kralicky — CHANGES_REQUESTED** [review]

File: pkg/resources/opnicluster/elastic/indices/ism.go [hunk 1]
```diff
@@ -7,19 +7,19 @@ import (
...
+	"github.com/opensearch-project/opensearch-go"
+	"github.com/opensearch-project/opensearch-go/opensearchapi"
 )
 
 type ISMApi struct {
-	*elasticsearch.Client
+	*opensearch.Client
 }
 
 func generatePath(name string) strings.Builder {
```
[the remark this answers is not in the document]
Comment (dbason) [reply pkg/resources/opnicluster/elastic/indices/ism.go:18]: Yeah the reason I did it this way was to mirror how it's done in the opensearch/elasticsearch clients.  I've actually opened an issue to have the ISM API included in the upstream opensearch client.

File: apis/v1beta1/opnicluster_types.go [hunk 2]
```diff
@@ -67,7 +67,7 @@ type OpniClusterSpec struct {
 type OpniClusterStatus struct {
 	Conditions        []string         `json:"conditions,omitempty"`
 	State             OpniClusterState `json:"state,omitempty"`
-	IndexState        OpniClusterState `json:"indexState,omitempty"`
```
[the remark this answers is not in the document]
Comment (dbason) [reply apis/v1beta1/opnicluster_types.go:70]: That field has been moved to the Opensearch Status struct

File: pkg/resources/opnicluster/opnicluster.go [hunk 3]
```diff
@@ -167,6 +198,54 @@ func (r *Reconciler) Reconcile() (retResult *reconcile.Result, retErr error) {
 	return
 }
 
+func (r *Reconciler) ReconcileElasticUpgrade() (retResult *reconcile.Result, retErr error) {
+	lg := log.FromContext(r.ctx)
+	if r.opniCluster.Status.OpensearchState.Version == nil || *r.opniCluster.Status.OpensearchState.Version == r.opniCluster.Spec.Elastic.Version {
```
[the remark this answers is not in the document]
Comment (dbason) [reply pkg/resources/opnicluster/opnicluster.go:203]: If the version is unset we don't want to requeue though, we want to skip to the next reconciler.  An unset version indicates it's a new Opensearch cluster and won't need an upgrade so the logic is exactly the same.

File: pkg/resources/opnicluster/elastic/indices/ism.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (kralicky) [reply pkg/resources/opnicluster/elastic/indices/ism.go:18]: cool - no change needed here

File: pkg/resources/opnicluster/opnicluster.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (kralicky) [reply pkg/resources/opnicluster/opnicluster.go:207]: sounds good

**Review by kralicky — APPROVED** [review]

**Pull request merged**